Respond with ICMP reply for traffic to services without backends

dylandreimerink opened the pull request:
So far we have been dropping packets meant for services which do not have endpoints/backends. This causes clients to needlessly wait for replies and to retry sending traffic. This PR adds the ability to send back an ICMP or ICMPv6 reply with Destination unreachable (type 3) + Port unreachable (code 3) whenever this happens. 

This behavior is controllable via a new `--service-no-backend-response` flag, which defaults to `reject` so we match expected behavior by default. It can also be set to `drop` to preserve the existing behavior in case that was desired.

This new behavior works for both North/South traffic entering a node and East/West traffic responding to a request from a pod within the cluster.

Fixes: #10002

```release-note
Respond with ICMP reply for traffic to services without backends
```

dylandreimerink committed:
option: Add `--service-no-backend-response` option

This commit adds a new flag/option/setting to control the behavior
of the datapath when we receive a packet for a service that has no
backends. We default this behavior to `reject` to match the default
behavior of kube-proxy.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Differentiate `DROP_NO_SERVICE` and `DROP_UNSUPP_SERVICE_PROTO`

So far we have been using `DROP_NO_SERVICE` when a service has no
backends as well as when traffic with an unsupported protocol is
sent to any service. This commit differentiates between the two
cases by introducing `DROP_UNSUPP_SERVICE_PROTO` for the latter.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Allow uneven lengths in `__bpf_memzero`

The `__bpf_memzero` only supported even lengths and `1`. This commit
fixes the shortcoming.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Send ICMP unreachable when no service backends are available

Before this commit we would always drop packets if we were unable to
find a backend for a service. This causes a client to retry multiple
times until a timeout is reached. This is not expected behavior since
kube-proxy and other CNI's will send an ICMP reply.

If enabled, we can now reply with an ICMP port unreachable message when
no backend is available for a service. This is implemented with a
separate tail call since crating the ICMP message can be quite complex.

This functionality is implemented for both IPv4 and IPv6. For IPv4
traffic the received packet is truncated to 64 bytes and  included in
the ICMP response. For IPv6 traffic up to 1232 bytes of the original
message is included in the ICMPv6 response. This matches the behavior
of kube-proxy.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf,tests: Add tests for the ICMP response for services without backend

This commit tests the functionality added in the previous commit. We
test that North/South traffic gets a response via the nodeport path and
that East/West traffic gets a response via the lxc path.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
complexity-tests: Add no service backend response to complexity tests

Ensure that the no service tail calls are not to complex, especially
for older kernels with the 4k instruction limit in combination with the
ICMPv6 checksum calculation.

`SERVICE_NO_BACKEND_RESPONSE=1` to {4,5,6}.txt for all kernels
expect 4.19 since this feature requires at least 5.2.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
test,l4lb: Change expected error in L4LB test

The current test expects curl to time out when backend weight is set to
0 because the drop packet behavior. With the new ICMP reply in such
cases we expect error code 7 not 28.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
probes: Added probe for bpf_skb_adjust_room + `BPF_ADJ_ROOM_MAC` mode

The --service-no-backend-response=reject feature requires the use of
`bpf_skb_adjust_room` with the `BPF_ADJ_ROOM_MAC` mode to make room to
add the outer IP + ICMP header. However this mode is only available
after v5.2, so this commit adds a probe to check for the availability
and will fall back to --service-no-backend-response=drop for kernels
that do not support it.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
test,conformance-ingress: Change sanity check to deal with ICMP reply

The current ingress conformance test adds an ingress and then does a
curl to confirm it works. In the past, Cilium would have dropped the
request packet until the datapath was setup. This silent dropping causes
curl to retry for 15 seconds before giving up.

With the ICMP reply however, curl gets an immediate response and will
give up immediately. This causes the test to fail. So this commit
adds manual retry logic and delays in the test script.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

lmb commented on test/l4lb/test.sh:
```diff
@@ -135,11 +135,11 @@ docker exec -t lb-node docker exec -t cilium-lb \
     curl -o /dev/null -m 0.5 "${LB_VIP}:80"
-    # code 28 - Operation timeout
-    if [ ! "$?" -eq 28 ]; then
+    # code 7 - Failed to connect ... : No route to host
```
Won't this break bisection between the commit where you change behaviour and this one?

dylandreimerink replied:
It will if you run CI while bisecting. I believe we expect the code to compile between every commit, not for the full CI suite to pass for every commit. That is what https://github.com/cilium/cilium/actions/runs/6197972695/job/16827550943 checks at least.

lmb commented on bpf/complexity-tests/419/bpf_lxc/4.txt:
```diff
@@ -24,3 +24,4 @@
 -DENABLE_IPV4_FRAGMENTS=1
 -DENCAP_IFINDEX=1
 -DTUNNEL_MODE=1
+-DSERVICE_NO_BACKEND_RESPONSE=1
```
What logic did you use to choose 0 vs 1? It'd be nice if you could document your thinking in the commit message.

lmb commented on bpf/lib/lb.h:
```diff
@@ -1207,7 +1207,7 @@ lb4_extract_tuple(struct __ctx_buff *ctx, struct iphdr *ip4, int l3_off, int *l4
 			return ret;
 		return 0;
 	case IPPROTO_ICMP:
```
Not related to this PR: does this mean we don't support forwarding ICMP to backends?

joestringer replied:
Yeah, k8s services operate on a specific port. You can hence have many services on the same IP on different ports. Depending on the ICMP message type there may not be a port at all, so how do you pick the right service?

lmb commented on bpf/lib/nodeport.h:
```diff
@@ -1322,6 +1322,14 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 				&key, &tuple, svc, &ct_state_new,
 				skip_l3_xlate, ext_err);
+
+#if SERVICE_NO_BACKEND_RESPONSE == 1
```
Why is does this compare against 1? Is there a comparison against 0 somewhere? From what I can tell the convention seems to be `if defined()` or similar instead of comparing the value.

dylandreimerink replied:
Yea, that is my bad. For a while it looked like we might need multiple fallback modes so that suck around, will update it.

lmb commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	/* Unrolled loop to calculate the checksum of the ICMP sample
+	 * Done manually because the compiler refuses with #pragma unroll
+	 */
+	csum = 0;
```
I'm really uneasy about this. Sticking it in the middle of the function makes it harder to test (is there a test somewhere?)

lmb commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	ip4->tos = tos;
+	ip4->tot_len = bpf_htons(sizeof(struct iphdr) + sizeof(struct icmphdr) +
+		       (__u16)sample_len);
+	ip4->id = (__be16)get_prandom_u32();
```
Controversial choice :D Can't we just set it to 0? The packet isn't fragmented after all. https://datatracker.ietf.org/doc/html/rfc6864

dylandreimerink replied:
Oh. Honestly never dove that deep into the RFCs. I have never seen 0 ids on PCAPs before so I figured setting wasn't optional :grin:. But it indeed seems like we can.

lmb commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	/* Write reversed icmp header */
+	icmphdr = data + sizeof(struct ethhdr) + sizeof(struct iphdr);
+	icmphdr->type = ICMP_DEST_UNREACH;
+	icmphdr->code = ICMP_HOST_UNREACH;
```
Have you checked what the consequences of going between host unreachable and port unreachable are? As in, how do network stacks react to this? Do they cache it?

I would be more comfortable with sticking to what kube proxy does.

asauber replied:
My quick thought here is that a Service K8s resource is defined as being associated with a port, so it would be important to NOT use host unreachable, as the IP is likely to be shared among multiple services and/or reused within some cache interval.

dylandreimerink replied:
That is a good point, I don't know if this is cached somehow. Agreed, lets stick to port unreachable for now since we know that works for others.

lmb commented on bpf/tests/tc_lxc_lb4_no_backend.c:
```diff
@@ -0,0 +1,140 @@
+
+	assert(l4->type == ICMP_DEST_UNREACH);
+	assert(l4->code == ICMP_HOST_UNREACH);
+	assert(l4->checksum == bpf_htons(0x7992));
```
How did you come up with this magic number?

dylandreimerink replied:
I dumped the context of the crafted packet with the DUMP_CTX option in the test runner, imported that into wireshark which then tells you if the checksum is correct and if not what it should be. I will add a comment to explain that.

lmb commented on bpf/tests/tc_lxc_lb6_no_backend.c:
```diff
@@ -0,0 +1,147 @@
+
+	assert(l4->icmp6_type == ICMPV6_DEST_UNREACH);
+	assert(l4->icmp6_code == ICMPV6_ADDR_UNREACH);
+	assert(l4->icmp6_cksum == bpf_htons(0x7da9));
```
Dito.

lmb reviewed (dismissed):
I'd really like to see and end to end test for this (sorry :see_no_evil:) ICMP and checksumming is especially finicky, and I'd like to make sure that we really get the expected reaction from a real network stack. It would also be great if the checksumming was it's own helper that was tested with more than three packets.

At my previous job we bungled the same ICMP checksumming code and would generate an invalid checksum with a probability of 1/1^16 or so.

joestringer commented on test/nat46x64/test.sh:
```diff
@@ -151,6 +151,9 @@ for i in $(seq 1 10); do
 done
 
+# Give the agent some time to persist service update to the datapath
+sleep 5
```
Please avoid sleeps in tests wherever possible. This is the cause of flakiness all over the place. Maybe query for presence of the service/backend instead on a regular period and fail out if it never shows up?

dylandreimerink replied:
Yea, might be able to use the `cilium bpf ...` commands to check when the service is actually "ready"

dylandreimerink replied:
So long story short, checking with `cilium bpf ...` commands did not work. My new hack is to allow a few connections to fail initially until it starts working with delays so we don't have a "dumb" sleep just there.

joestringer commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	/* Redirect ICMP to the interface we received it on. */
+	cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
+			   ctx_get_ifindex(ctx));
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
```
How do we ratelimit ICMP? [RFC4443](https://datatracker.ietf.org/doc/html/rfc4443) 2.4.f requires implementations to limit how many ICMPv6 messages that a node generates.

dylandreimerink replied:
Well we currently don't since I glossed over that part of the spec :fearful:. Any suggestions on how to add that? 

First instinct is to do so in BPF, add a map to track messages and just drop the incoming packet if we hit the quota. Though I suspect someone will insist on making such a mechanism generic and reusable to warrant the cost of introducing a new map.

Second thought is to re-use parts of the bandwidth manager. Which would mean that if we want this behavior to be the default (as it should be for k8s conformance) that bandwidth manager is always enabled. We might be able to just re-use the TC setup in the background so we can use it to rate-limit, without having to enable the whole feature. But it sounds ticky in any case.

I would urge a follow up PR for this works since it seems significant to add, even though that means we technically are not RFC compliant for now. I would argue that its of less importance in this code path anyway since we specifically deal with replies for services without backends, which is a very specific scenario as apposed to something like a echo reply.

joestringer replied:
I'm not quite sure about the "map to track messages" approach, the standard just suggests a token bucket implementation. Bandwidth manager seems like it may have more in common.

As a general principle, I would be more comfortable with merging such functionality into a branch and potentially breaking up individual tasks into PRs to target that branch, then only merge the branch into main when the feature is complete / standards-compliant. I worry that it's too easy to say "well that's a big item so we'll do it later" and then later becomes ~forever, and Cilium ends up releasing a version that violates well-established standards. This is part of the spec for the feature that's being added in this PR.

dylandreimerink replied:
> I worry that it's too easy to say "well that's a big item so we'll do it later" 

I personally don't mind putting it all into one PR, but I have been told that smaller PRs are easier to review hence the initial suggestion. :grin: 

> I would be more comfortable with merging such functionality into a branch and potentially breaking up individual tasks into PRs to target that branch

If we can merge this into a `feature/...` branch with reviewed PRs and then merge the whole branch into main as a whole that would be optimal in my opinion.

> I'm not quite sure about the "map to track messages" approach, the standard just suggests a token bucket implementation. Bandwidth manager seems like it may have more in common.

I meant implementing something like a token bucket algorithm with a map as storage. But I agree that reusing the bandwidth manager would be better than making something new at this point.

> [RFC4443](https://datatracker.ietf.org/doc/html/rfc4443) 2.4.f requires implementations to limit how many ICMPv6 messages

Motivation behind this PR is reports of users hitting this issue. Would it be acceptable to rip out the IPv6 path since its only the ICMPv6 spec that requires rate limiting? That way we can ship something in a reasonable amount of time, given I expect the bandwidth manager route to take significant time

I realize that is another way of saying "well that's a big item so we'll do it later", but its for a good cause :angel:

joestringer replied:
I've created `feature/main/svc-icmp-response`, you should be able to submit PRs against that branch. The goal should be to ensure that merges into that branch match the expectations for merging into `main`. Periodically if necessary we can rebase it (I think you should have the ability to do this, but let me know if not). Once we think the feature is complete we can submit a PR to pull that branch into main.

> Motivation behind this PR is reports of users hitting this issue. Would it be acceptable to rip out the IPv6 path since its only the ICMPv6 spec that requires rate limiting? That way we can ship something in a reasonable amount of time, given I expect the bandwidth manager route to take significant time

This has been the behaviour for _years_. This is a brand new feature being submitted half way through the v1.15 development cycle. I don't see the urgency. I'll note that arguments like this are a common cause of tech debt and also a big reason why various features don't work with IPv6. By all means we can have a discussion with the relevant sig-datapath folks to set the expectations according to the shared community quality bar, but I don't really buy the argument why corners must be cut here.

joestringer commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	pseudo_header.fields.next_header = IPPROTO_ICMPV6;
+
+	#pragma unroll
+	for (i = 0; i < 20; i++)
```
What's `20` here?

dylandreimerink replied:
Its the size of the words array `__u16 words[20];` in the pseudo header, but I think I can exchange that for `sizeof(pseudo_header.words)/sizeof(__u16)` to make it more clear

joestringer commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+
+csumend :
+
+	/* We need to insert a IPv4 and ICMP header before the original packet.
```
Why IPv4 in the IPv6 function?

joestringer commented on bpf/lib/lb.h:
```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+#if __ctx_is == __ctx_xdp
+	xdp_adjust_head(ctx, 0 - (int)(sizeof(struct ipv6hdr) + sizeof(struct icmp6hdr)));
+#else
+	skb_adjust_room(ctx, sizeof(struct ipv6hdr) + sizeof(struct icmp6hdr), BPF_ADJ_ROOM_MAC, 0);
```
No error checking?

Also curious why this isn't using `ctx_adjust_hroom`? Is there an import issue or something that prevents reuse of that helper?

dylandreimerink replied:
> No error checking?

Yea, I don't really have a recourse if these error, and I don't think it warrants creating a special error code. But I can always do an early return.

> Also curious why this isn't using ctx_adjust_hroom?

Well, initially I used the `bpf_skb_change_head` helper, but it doesn't for TC until v5.8, I only later switched. Though looking at the possibility now, I use the `BPF_ADJ_ROOM_MAC` mode which isn't implemented in our XDP implementation of the helper, the code currently asserts the `BPF_ADJ_ROOM_NET` mode. So we would have to add that as well (and the PR is already chunky).

joestringer reviewed (commented):
I got a bit curious about the checksum implementation and I got engrossed in digging around in some protocol details. Forgive the pass-by review :sweat_smile: . Not a comprehensive review by any means, I just noticed a few things. I note a few points where you've clearly been implementing to the RFC specs, but there was also some finer points of the RFCs not yet addressed as far as I could tell.

asauber commented on .github/workflows/conformance-ingress.yaml:
```diff
@@ -242,8 +242,25 @@ jobs:
       - name: Run Sanity check (external)
         timeout-minutes: 5
         run: |
+          set +e
```
Adding errexit here will cause this test to exit upon the curl failing, and not reach the `if [ $? -ne 0 ];` lines.

dylandreimerink replied:
I believe this does the exact opposite, by default these scripts are run with `set -e` which causes an exit of the whole script on error, I had to add this line so I could use the `if [ $? -ne 0 ];` checks

asauber commented on install/kubernetes/cilium/README.md:
```diff
@@ -740,6 +740,7 @@ contributors across the globe, there is almost always someone available to help.
 | serviceAccounts.clustermeshcertgen | object | `{"annotations":{},"automount":true,"create":true,"name":"clustermesh-apiserver-generate-certs"}` | Clustermeshcertgen is used if clustermesh.apiserver.tls.auto.method=cronJob |
 | serviceAccounts.hubblecertgen | object | `{"annotations":{},"automount":true,"create":true,"name":"hubble-generate-certs"}` | Hubblecertgen is used if hubble.tls.auto.method=cronJob |
 | serviceAccounts.nodeinit.enabled | bool | `false` | Enabled is temporary until https://github.com/cilium/cilium-cli/issues/1396 is implemented. Cilium CLI doesn't create the SAs for node-init, thus the workaround. Helm is not affected by this issue. Name and automount can be configured, if enabled is set to true. Otherwise, they are ignored. Enabled can be removed once the issue is fixed. Ciliu ...
+| serviceNoBackendResponse | string | `"reject"` | Configure what the response should be to traffic for a service without backends. Possible values:  - reject (default)  - drop |
```
A note here along the lines of "earlier kernels (< 5.10[?]) do not support the 'reject' mode", would be helpful. An upgrade note along these lines I think is also required.

asauber commented on .github/workflows/conformance-ingress.yaml:
```diff
@@ -242,8 +242,25 @@ jobs:
+          # flag only retries on connection timeouts.
+          curl -s -v --connect-timeout 5 --max-time 20 --fail -- http://"$lb"
+          if [ $? -ne 0 ]; then
+            sleep 5s
```
We want to avoid sleeps in tests, and this could be a loop with a counter rather than nested, but I think both are _okay_ here.

joestringer replied:
FWIW the danger is that if we have a bug that causes LB decisions to be flaky or inconsistent rather than moving from reliably not working to reliably working, then this style of testing will hide such flaky production behaviour.

asauber reviewed (changes requested):
Thanks for switching to port unreachable.

A few more minor comments.

dylandreimerink committed:
bpf: Add ratelimiting feature and use it for ICMPv6 responses

This commit adds token bucket ratelimiting to the datapath. It is
implemented purely in BPF. A new map is added to keep track of the
buckets. A bucket can be keyed on anything, though since ICMPv6 is
currently the only user it is keyed on ifindex. The value is the current
amounts of tokens in the bucket and the last time we added tokens to the
bucket.

For every event we check if there is at least 1 token left in the bucket
if so, we decrement the token count and continue, if not we execute the
rate limiting action. Typically a timer would add new tokens into the
bucket, in our case we keep track of the last time we added tokens and
calculate how many tokens we should have added since then before we
do the token check.

This implements a burstable rate limiting mechanism. The burst size
and token refil is configurable. For ICMPv6 it is currently set to
100 replies per second with a burst size of 1000.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki commented on bpf/lib/lb.h:
```diff
@@ -1917,13 +1918,24 @@ int __tail_no_service_ipv6(struct __ctx_buff *ctx)
 	struct in6_addr saddr;
 	struct in6_addr daddr;
+	struct ratelimit_key rkey = {};
+	/* Rate limit to 100 ICMPv6 repies per second, burtable to 1000 responses/s */
```
typos

kaworu reviewed (approved):
Hubble changes LGTM

sayboras reviewed (approved):
LGTM on helm changes

asauber commented on .github/workflows/conformance-ingress.yaml:
```diff
@@ -243,7 +243,21 @@ jobs:
         run: |
           lb=$(kubectl get ingress basic-ingress -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
-          curl -s -v --connect-timeout 5 --max-time 20 --retry 3 --fail -- http://"$lb"
+          
```
Instead of adding this logic, have you tried using cURL's `--retry-all-errors` flag? There is also `--retry-connrefused`.

It could simplify things here.

dylandreimerink replied:
Awesome, didn't know about this flag. Works according to CI, so going with this! Thx

asauber reviewed (approved):
Another suggestion for the integration test. Not blocking.

Token bucket implementation looks clean. Thanks again again for port unreachable.

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -484,6 +484,57 @@ func HaveOuterSourceIPSupport() (err error) {
+	// This is a Eth + IPv4 + UDP + data packet. The helper relies on a valid packed being passed in
+	// since it wants to know offsets of the different layers.
+	pkt := []byte{
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0e, 0xf5, 0x16, 0x3d, 0x6b, 0xab, 0x08, 0x00, 0x45, 0x00,
```
Generate this using `gopacket` :pray:

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -484,6 +487,78 @@ func HaveOuterSourceIPSupport() (err error) {
+	}
+	defer prog.Close()
+
+	// This is a Eth + IPv4 + UDP + data packet. The helper relies on a valid packed being passed in
```
Nit (sorry)
```suggestion
	// This is a Eth + IPv4 + UDP + data packet. The helper relies on a valid packet being passed in
```

dylandreimerink replied:
(sigh)

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -484,6 +487,78 @@ func HaveOuterSourceIPSupport() (err error) {
+		}
+	}()
+
+	if err := rlimit.RemoveMemlock(); err != nil {
```
This can likely be removed, same with the one in `HaveOuterSourceIPSupport()` above.

gentoo-root commented on bpf/lib/ratelimit.h:
```diff
@@ -0,0 +1,76 @@
+	since_last_topup = ktime_get_ns() - value->last_topup;
+	if (since_last_topup > settings->topup_interval_ns) {
+		/* Add tokens of every missed interval */
+		value->tokens += (since_last_topup / settings->topup_interval_ns) *
```
Rounding here could skip intervals. For example, it this function is called at 0 s, 1.5 s and 3 s, it will add 1000 tokens at 0 s, another 1000 tokens at 1.5 s, and yet another 1000 tokens at 3 s = in total 3000 tokens. If, however, this function was called at 0 s, 1 s, 2 s and 3 s, it would add 1000 tokens at each call = in total 4000 tokens over the same time period.

dylandreimerink replied:
Right. So a better way to keep track would be:
```
long cur_time = ktime_get_ns();
[...]
long intervals = since_last_topup / settings->topup_interval_ns;
long remainder = since_last_topup % settings->topup_interval_ns;
value->last_topup = cur_time - remainder;
```

So a 1.5s, we set `last_topup` to `1s` instead of `1.5s`. So at the 3s mark we would add 2000 instead of 1000.

Is that correct?

gentoo-root replied:
Looks correct to me.

gentoo-root commented on bpf/lib/ratelimit.h:
```diff
@@ -0,0 +1,76 @@
+	if (!value) {
+		new_value.last_topup = ktime_get_ns();
+		new_value.tokens = settings->tokens_per_topup - 1;
+		ret = map_update_elem(&RATELIMIT_MAP, key, &new_value, BPF_ANY);
```
This lookup-and-update is racy if called from two CPUs. Do we care?

dylandreimerink replied:
I considered this. It would mean that the ratelimit isn't 100% accurate, letting more traffic through than the limit. To fix that we would need to use atomics which are slow. I thought the performance is more important in this given situation. Perhaps we should add this to the comments in case other wonder.

gentoo-root commented on bpf/lib/ratelimit.h:
```diff
@@ -0,0 +1,76 @@
+		/* Add tokens of every missed interval */
+		value->tokens += (since_last_topup / settings->topup_interval_ns) *
+						  settings->tokens_per_topup;
+		value->last_topup = ktime_get_ns();
```
We should reuse the ktime_get_ns() value fetched above, otherwise it's another source of inexactness, although only a tiny one.

gentoo-root reviewed (commented):
I'm too late for the party, but I've got some comments on the TBF implementation :sweat_smile:

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1741,8 +1746,357 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+
+	/* Redirect ICMP to the interface we received it on. */
+	cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
+			   ctx_get_ifindex(ctx));
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
+}
```
I believe this needs an `edt_set_aggregate(ctx, 0)`, to prevent false-positives in `to-netdev`'s Bandwidth-Manager code.

dylandreimerink replied:
That is a good point, I have not been able to check the interaction with the bandwidth manager. But I suspect you are right. Will have to do a followup.

julianwiedmann replied:
I'm playing around in https://github.com/cilium/cilium/pull/33046.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1741,8 +1746,357 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
+			   ctx_get_ifindex(ctx));
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
+}
```
When this ICMP packet hits `to-netdev`, did you check how it interacts with the [SNAT engine](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/bpf_host.c#L1423)?

I would expect that it gets [dropped](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/lib/nat.h#L778), whenever the addressed service IP equals [IPV4_MASQUERADE](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/lib/nat.h#L526).

dylandreimerink replied:
> did you check how it interacts with the [SNAT engine](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/bpf_host.c#L1423)?

No, I did not. 

> I would expect that it gets [dropped](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/lib/nat.h#L778), whenever the addressed service IP equals [IPV4_MASQUERADE](https://github.com/cilium/cilium/blob/2daca4c3935a43dac72044c7ca3e17ac57af2617/bpf/lib/nat.h#L526).

Right, which would be in the case of a node port service?

julianwiedmann replied:
> which would be in the case of a node port service?

Correct. That's a scenario we want to support, right?

dylandreimerink replied:
Yes, I think so. How does host traffic normally deal with this? Additionally, I see where have a marker to skip SNAT `ctx_snat_done_set(ctx);` would calling it before doing the redirect help?

It has been a while since I looked in depth at the SNAT path.

julianwiedmann replied:
> Yes, I think so. How does host traffic normally deal with this?

It gets dropped :). We only support a limited set of ICMP types. I hope we can extend this as needed.

> Additionally, I see where have a marker to skip SNAT `ctx_snat_done_set(ctx);` would calling it before doing the redirect help?

Yep, I had the same thought. I believe that would fit as work-around (and would even allow you to skip the HostFW in `to-netdev` ... that's another aspect we didn't consider yet in this PR). Long-term it would be best to teach the SNAT engine about this ICMP type.

julianwiedmann replied:
Copied to https://github.com/cilium/cilium/issues/33041.

julianwiedmann reviewed (commented):
Two comments below - better late than never :).

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -108,6 +108,14 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
 		ret = lb4_local(get_ct_map4(&tuple), ctx, ipv4_is_fragment(ip4),
 				ETH_HLEN, l4_off, &key, &tuple, svc, &ct_state_new,
 				has_l4_header, false, &cluster_id, ext_err);
+
+#ifdef SERVICE_NO_BACKEND_RESPONSE
+		if (ret == DROP_NO_SERVICE) {
+			ep_tail_call(ctx, CILIUM_CALL_IPV4_NO_SERVICE);
+			return DROP_MISSED_TAIL_CALL;
+		}
+#endif
+
```
With ep-routes enabled, the ICMP packet should now pass through `to-container` on the way back into the pod. Would we thus require an ingress network policy change to allow this traffic?

This feels very similar to the topic of avoiding policy for service loopback replies ...

dylandreimerink replied:
A CT entry should be created for the outgoing connection to the service, so when doing policy checking the ICMP reply should be flagged as return traffic an thus not subject to any ingress policy.

julianwiedmann replied:
Unfortunately [lb4_local()](https://github.com/cilium/cilium/blob/4400169ba429babe013419b363d8c00e6872ff0f/bpf/lib/lb.h#L1579) currently doesn't create the RELATED entry (note the `NULL` for `map_related`). So I don't think there's any CT entry in place that would allow such ICMP traffic to pass through network policy enforcement.

dylandreimerink replied:
Right, but isn't the CT entry created [here](https://github.com/cilium/cilium/blob/4400169ba429babe013419b363d8c00e6872ff0f/bpf/bpf_lxc.c#L959) before we get to the LB stage?

dylandreimerink replied:
I see that depends on `ENABLE_PER_PACKET_LB` being enabled or not

julianwiedmann replied:
> Right, but isn't the CT entry created [here](https://github.com/cilium/cilium/blob/4400169ba429babe013419b363d8c00e6872ff0f/bpf/bpf_lxc.c#L959) before we get to the LB stage?

Nope, that part is only reached *after* selecting the backend (this CT entry tracks the client -> backend connection).

julianwiedmann replied:
Copied to https://github.com/cilium/cilium/issues/33041.

ti-mo commented on bpf/lib/common.h:
```diff
@@ -597,7 +599,7 @@ enum {
 #define DROP_INVALID_EXTHDR	-156
 #define DROP_FRAG_NOSUPPORT	-157
 #define DROP_NO_SERVICE		-158
-#define DROP_UNUSED8		-159 /* unused */
```
@dylandreimerink This drop reason wasn't added in `flow.proto` and `drop.go`.

Under normal circumstances, we shouldn't reuse any of these, since renaming a proto field/type causes a backwards-incompatible change. We're lucky that in this case, drop reason `159` is actually missing from the proto as well as from drop.go. :sweat_smile: 

```
    SERVICE_BACKEND_NOT_FOUND = 158;
    NO_TUNNEL_OR_ENCAPSULATION_ENDPOINT = 160;
```

In any case, I'm marking all unused ones as deprecated in https://github.com/cilium/cilium/pull/29482.

cc @rolinh

julianwiedmann commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2276,6 +2276,13 @@ routingMode: ""
 # @default -- Port 8472 for VXLAN, Port 6081 for Geneve
 tunnelPort: 0
 
+# -- Configure what the response should be to traffic for a service without backends.
+# "reject" only works on kernels >= 5.10, on lower kernels we fallback to "drop".
+# Possible values:
+#  - reject (default)
```
@dylandreimerink is the >= 5.10 requirement correct? We later say that the required BPF feature (`BPF_ADJ_ROOM_MAC` support in `bpf_skb_adjust_room()`) exists since kernel 5.2.

dylandreimerink replied:
Yes, I believe you are correct. That should be 5.2, I am not sure how 5.10 ended up there.

julianwiedmann replied:
ack, I glanced at a 5.4 sysdump and it looks all golden there. No error message from the probe.

I'll open a PR to just remove this on v1.16+, given all supported kernels should be providing the option.

julianwiedmann replied:
https://github.com/cilium/cilium/pull/35734
